File: fsharp/valentines-day/ValentinesDay.fs
module ValentinesDay

type Approval = Yes | No | Maybe

type Cuisine = Korean | Turkish

type Genre = Crime | Horror | Romance | Thriller

type Activity = BoardGame | Chill | Movie of Genre | Restaurant of Cuisine | Walk of int

let rateActivity activity = 
    match activity with
    | Movie(Romance) | Restaurant(Korean) -> Yes
    | Restaurant(Turkish) -> Maybe
    | Walk(distance) when distance < 3 -> Yes
    | Walk(distance) when distance < 5 -> Maybe
    | _ -> No

File: haskell/valentines-day/src/ValentinesDay.hs
module ValentinesDay where

data Approval
    = Yes
    | No
    | Maybe

data Cuisine
    = Korean
    | Turkish

data Genre
    = Crime
    | Horror
    | Romance
    | Thriller

data Activity
    = BoardGame
    | Chill
    | Movie Genre
    | Restaurant Cuisine
    | Walk Int

rateActivity :: Activity -> Approval
rateActivity (Movie Romance) = Yes
rateActivity (Restaurant Korean) = Yes
rateActivity (Restaurant Turkish) = Maybe
rateActivity (Walk distance) 
    | distance < 3 = Yes
    | distance <= 5 = Maybe
    | otherwise = No
rateActivity _ = No
